### ISSUE TITLE
[BUG] sensor.eforsyning_energy_used suddely jumped on 31/12-2022

### ISSUE THREAD

**Issue by Symbiot78** [kpoppel/homeassistant-eforsyning#48]
hi 

the sensor for energy used suddenly jumped from 38 to 106 in my integration.
I cannot figure out where something went wrong.
looking in the eForsyning App or website I do not see this number anywhere nor can I see which datapoint would supply this number.

The exp sensor shows sensor.eforsyning_energy_exp_used shows 57.

I'd like to find the issue but not sure how to proceed.. please advice.

**Comment by Symbiot78** [comment]
same goes for the water used & exp used.

**Comment by kpoppel** [comment]
The integration only formats data from the eForsyning API in the format Home Assistant likes, and zeroes out data which are verifiable wrong, like cooling or return temperature of 85 or more degrees which is caused by read errors on the meter.

There are different sensors in play, some show daily use and expected use, and some the total used and expected used.  I hope the names are clear enough for this difference.
Could the sensor you are looking at be the total used sensors which would jump up day to day from zero after New Year?

I my web-eForsyning I can see non totals data from 2023 on the year and month views, but the daily view does show something.
I am not actually sure the integration handles the case where these two datasets are unavailable, I will check.

**Comment by Symbiot78** [comment]
this is the energy end sensor.. strange behavior between the 2 dates..

![2023-01-07_17-06-14](https://user-images.githubusercontent.com/3314249/211159879-186c0ee5-6638-45fe-af1e-7116dfd19641.png)

**Comment by Symbiot78** [comment]
just checked my meter. 25834 

so from 31/12 until today I've used 193 Kwh -which is 27Kwh pr. day average
If I tage the number from 30/12 and subtract the value from 24/12 I get 249Kwh which averages 34Kwh pr. day.


these are the 2 charts that are making me question the numbers.. water and kwh

![2023-01-07_17-19-26](https://user-images.githubusercontent.com/3314249/211160394-43217ec2-59ae-4e59-9743-ca92f2e247fd.png)


I wonder if it's ""just"" unexpected usage.. I wish I could see data from january already :(

**Comment by kpoppel** [comment]
Yes, that would raise an eyebrow or two :-)
It looks like I need to look at the code to handle the case where no month or year data is available, but day data is.
As mentioned in the web-ui I can see day data if I explicitly select it, but there is no "2023" bar yet - even though it should be there since daily data is okay already (as expected).

Stay tuned, maybe I can fish out this so January can get to look okay :-)

**Comment by kpoppel** [comment]
Okay, the trouble is that there is this parameter in the API call which designates which year is relevant.  It is important because some have years from July to June, others from October to September, and others from January to December (like me).  Those with skewed years probably won't see any of this weirdness.

If I get data using the marker I can get from the API I cannot get January data, and monthly and yearly data is "empty".  If I use current year, those with skewed years will get in trouble.
To fix it I need to make a trial if currentYear != "API year" to see if data makes sense, then sort out the fact that totals don't exist yet.  Then, when the "API year" is updated, stop doing that.
Now that's going to be fun ;-)

**Comment by Symbiot78** [comment]
> Okay, the trouble is that there is this parameter in the API call which designates which year is relevant. It is important because some have years from July to June, others from October to September, and others from January to December (like me). Those with skewed years probably won't see any of this weirdness.
> 
> If I get data using the marker I can get from the API I cannot get January data, and monthly and yearly data is "empty". If I use current year, those with skewed years will get in trouble. To fix it I need to make a trial if currentYear != "API year" to see if data makes sense, then sort out the fact that totals don't exist yet. Then, when the "API year" is updated, stop doing that. Now that's going to be fun ;-)

Would it be an option to just put the start of year as a config patan for the user to input?

**Comment by Symbiot78** [comment]
If that would make things easier?

**Comment by kpoppel** [comment]
I have asked the developers of eForsyning and it is actually something which is controlled by the local heating company.  They have to update a year marker which can even be automated on their side.  I have made a workaround which tries to bridge the gap by optimistically getting the new year data and if that is valid, just go with that during the first month of the year.
I will make a pre-release of this while I monitor the data coming out.

**Comment by kpoppel** [comment]
Try pre-release v1.1.5 It seems to do the job for me at least.

**Comment by Symbiot78** [comment]
Hmm
I see no difference after upgrading to .15
Same spikes

**Comment by henrikbol** [comment]
Had the same issue (Vestfor). Before (v 1.1.4): Looked ikke this:
![Screenshot 2023-01-15 at 17 08 12](https://user-images.githubusercontent.com/44621678/212552700-592896df-6797-4a22-8852-beb8217e2bf2.png)
Now at v 1.2  It looks like this.
![Screenshot 2023-01-15 at 17 19 30](https://user-images.githubusercontent.com/44621678/212552884-f68ed6b7-fe47-4bec-931e-4f5a64cf3ee2.png)

 So I only see 2023 data on `eforsyning_energy_used` and `eforsyning_energy_exp_used`

And the exp-used is way off, but I assume that is from the API. On eForsyning my expected use is not updated for 2023.

**Comment by kpoppel** [comment]
Yes.  With this change data is opportunistically fetched during January for the correct year to compensate for a value the heating company needs to update.  This only works out for some sensors as others are updated from yearly data or budget prognosis.  It also means that before all data from 2022 were fetched, the upper picture, to the new data from the current year to be fetched - the history resets every year.
 If you want historical data, you need to setup something like "keep_days: <days>" or better use InfluxDB and Grafana or custom sensors to get it back.  Home Assistant is not effective when saving a lot of data in the SQLite database.

In any case it is impossible to show data which simply does not exist.  My own local deliverer wrote me back that accumulated data will not appear until February, meaning some sensors may simply work off 2022 numbers, meaning registering a constant at best.  My used vs. expected seems okay.  I have not seen a jump like from 84 kWh to 24 kWh

**Comment by smulle48** [comment]
I've looked through my debug log and found something interesting.

It looks like they have changed the api.

According to this plugin, the url to my data has always been:

https://apiapp.dff-edb.dk/aabenraa/api/getebrugerinfo?id=536f0e012e19bb12b748fb092dXXXXXX

But when I open the homepage and look at the data, it fetch the following url (where also all the 2023 data is included)

https://apiapp.dff-edb.dk/aabenraa/api/getebrugerinfo?id=4cc3c10137aa879cd140011359XXXXXX&unr=XXXXXXXXX&anr=2&inr=1

The same goes for:

https://apiapp.dff-edb.dk/aabenraa/api/getforbrug?id=4cc3c10137aa879cd140011359XXXXXX&unr=XXXXXXXXX&anr=2&inr=1
